server/matchmaker/algorithm/team_matchmaker.py

```diff
@@ -0,0 +1,286 @@
+        self.rating = rating_difference
+        self.content = content
+
+
+class UnevenTeamsException(Exception):
+    pass
+
+
+class NotEnoughPlayersException(Exception):
+    pass
+
+
+@with_logger
+class TeamMatchMaker(Matchmaker):
+    """
+    Matchmaker for teams of varied size. Untested for higher than 4v4 but it should work
+
+    Overview of the algorithm:
+    1. list all the parties in queue by their average rating.
+    2. Take a party and select the neighboring parties alternating between lower
+    and higher until you have 8 players. If the next party to select would leave
+    you with more than 8 players you skip that party and try the next one.
+    3. you now have a list of parties that will be in your potential game. Distribute
+     them into two teams.
+     Start with the easiest cases: one party makes a full team already or one party
+     has n-1 players so you only need to find the best fitting single player.
+     If that is not the case perform the karmarkar-karp algorithm to get a good approximation for a partition
+    4. add this game to a games list
+    5. repeat 2. to 4. for every party.
+    6. you now have a list of potential games with minimal rating variation and minimal rating imbalance.
+    7. remove all games with match quality below threshold then sort by quality descending
+    8. pick the first game from the game list and remove all other games that contain the same players
+    9. repeat until the list is empty
+
+    for the future:
+    Find combination of potential games that allows for maximal number of games to launch
+    Repeat 8. with an increasing amount of top games removed beforehand to get all possible game combinations
+    If there is more than one solution, pick the one with highest total game quality
+    Optimization: sort by game quality first, abort when you find a solution with
+    the full number of theoretically possible games (floor(playersInQueue/(teamSize*2)))
+    """
+
+    def find(self, searches: Iterable[Search]) -> List[Match]:
+        self._logger.debug("========= starting matching algorithm =========")
+
+        searches = SortedList(searches, key=lambda s: s.average_rating)
+        possible_games = []
+        for index, search in enumerate(searches):
+
+            self._logger.debug("building game for %s", repr(search))
+
+            try:
+                participants = self._pick_neighboring_players(searches, index)
+                match = self.make_teams(participants)
+                game = self.assign_game_quality(match)
+                possible_games.append(game)
+            except NotEnoughPlayersException:
+                self._logger.warning("Couldn't pick enough players for a full game. Skipping this game...")
+            except UnevenTeamsException:
+                self._logger.warning("Failed to assign even teams. Skipping this game...")
+
+        self._logger.debug(
+            "got %i games\n" % len(possible_games) + "\n".join(
```

Comment:
Use f-strings instead of `%` operator.:
```suggestion
            f"got {len(possible_games)} games\n\n".join(
```

Reply:
I thought you dislike the f-strings because they are less performant? Do you want me to change only this one specifically or all?

Reply:
Well, in general I am a fan of f-strings, but the logger library doesn't support that style of formatting in log messages afaik. Which means if you use f-strings in log calls the string formatting will happen eagerly even if the string is never printed. This means ever so slightly worse performance that you will never notice except possibly in very tight loops. But I still prefer the '%' way for "correctness" and consistency because we do it in all log statements in this project.

However, in this case you are eagerly formatting the string anyways because you want to print out many lines of output, so I'd just go ahead and use an f-string.

Now that I think of it though. There might be an API for checking what the log level of the logger is so you could do that and then put your for loop that prints a bunch of debug inside of an if statement.

```python3
if self._logger.isEnabledFor(logging.DEBUG):
    # Put as many debug printing loops as you want here
    # because they won't run in INFO logging mode
```